src/analyses/cfg_dominators.h

```diff
@@ -115,6 +246,29 @@ class cfg_dominators_templatet
 protected:
   void initialise(P &program);
   void fixedpoint(P &program);
+
+  /// Goal of this is to process the tree in a post-order fashion.
+  /// Full explanation here: https://en.wikipedia.org/wiki/Tree_traversal.
+  ///
+  /// Post-ordering is when you pick one path of the tree to walk down to its
+  /// end, number that node, walk up one, do the same with the other path,
+  /// then repeat. When you've already numbered both children of a node, you
+  /// number it, then walk up one. This then continues until you've walked the
+  /// entire tree.
+  void assign_postordering(std::size_t start_node_index);
+
+  /// Sort our CFG nodes in reverse post-order. Just gets every node that we've
+  /// processed (anything not NODE_NOT_VISITED) then order from highest to
+  /// lowest.
+  std::vector<T>
+  get_reverse_postordered_instructions(std::size_t entry_node_index) const;
+
+  /// Walks up the dominators of left/right nodes until it finds one that is
+  /// common to both sides. Used to work out a the common immediate dominator
+  /// when a node that has multiple incoming edges.
```

Comment:
Cite figure 3 in "A Simple, Fast Dominance Algorithm" by Cooper, Harvey and Kennedy.